src/protocols/rdp/channels/cliprdr.c

```diff
@@ -0,0 +1,601 @@
+               writer, &output, GUAC_RDP_CLIPBOARD_MAX_LENGTH);
+
+    CLIPRDR_FORMAT_DATA_RESPONSE data_response = {
+        .requestedFormatData = (BYTE*) start,
+        .dataLen = ((BYTE*) output) - start,
+        .msgFlags = CB_RESPONSE_OK
+    };
+
+    guac_client_log(clipboard->client, GUAC_LOG_TRACE, "CLIPRDR: Sending "
+            "format data response.");
+
+    return cliprdr->ClientFormatDataResponse(cliprdr, &data_response);
+
+}
+
+/**
+ * Callback invoked by the FreeRDP CLIPRDR plugin for received Format Data
+ * Response PDUs. The Format Data Response PDU is sent by the RDP server when
+ * fullfilling a request for clipboard data received via a Format Data Request
+ * PDU.
+ *
+ * @param cliprdr
+ *     The CliprdrClientContext structure used by FreeRDP to handle the CLIPRDR
+ *     channel for the current RDP session.
+ *
+ * @param format_data_response
+ *     The CLIPRDR_FORMAT_DATA_RESPONSE structure representing the Format Data
+ *     Response PDU that was received.
+ *
+ * @return
+ *     CHANNEL_RC_OK (zero) if the PDU was handled successfully, an error code
+ *     (non-zero) otherwise.
+ */
+static UINT guac_rdp_cliprdr_format_data_response(CliprdrClientContext* cliprdr,
+        const CLIPRDR_FORMAT_DATA_RESPONSE* format_data_response) {
+
+    /* FreeRDP-specific handlers for CLIPRDR are not assigned, and thus not
+     * callable, until after the relevant guac_rdp_clipboard structure is
+     * allocated and associated with the CliprdrClientContext */
+    guac_rdp_clipboard* clipboard = (guac_rdp_clipboard*) cliprdr->custom;
+    assert(clipboard != NULL);
+
+    guac_client_log(clipboard->client, GUAC_LOG_TRACE, "CLIPRDR: Received "
+            "format data response.");
+
+    char received_data[GUAC_RDP_CLIPBOARD_MAX_LENGTH];
+
+    guac_iconv_read* reader;
+    const char* input = (char*) format_data_response->requestedFormatData;
+    char* output = received_data;
+
+    /* Find correct source encoding */
+    switch (clipboard->requested_format) {
+
+        /* Non-Unicode (Windows CP-1252) */
+        case CF_TEXT:
+            reader = GUAC_READ_CP1252;
+            break;
+
+        /* Unicode (UTF-16) */
+        case CF_UNICODETEXT:
+            reader = GUAC_READ_UTF16;
+            break;
+
+        /* If the format ID stored within the guac_rdp_clipboard structure is actually
+         * not supported here, then something has been implemented incorrectly.
+         * Either incorrect values are (somehow) being stored, or support for
+         * the format indicated by that value is incomplete and must be added
+         * here. The values which may be stored within requested_format are
+         * completely within our control. */
+        default:
+            guac_client_log(clipboard->client, GUAC_LOG_DEBUG, "Requested "
+                    "clipboard data in unsupported format (0x%X).",
+                    clipboard->requested_format);
+            return CHANNEL_RC_OK;
```

Comment:
Yes. FreeRDP handles error return codes similarly to the way we do, with such errors being treated as fatal (or at least near-fatal). In this case, I'm not sure whether the entire RDP connection would terminate, but returning a non-zero code here would result in the message processing thread for CLIPRDR being shut down.

The root of the relevant code flow is the thread body of FreeRDP's CLIPRDR handling:

https://github.com/FreeRDP/FreeRDP/blob/32bf768652a1a6d6872453ff5c84ab97bedaaa00/channels/cliprdr/client/cliprdr_main.c#L891-L935

In a nutshell:

1. Loading FreeRDP's "cliprdr" plugin results in the above thread starting. It handles messages received along the CLIPRDR channel, delegating processing of those messages to `cliprdr_order_recv()`, an internal function.
2. In the case of a Format Data Response PDU, `cliprdr_order_recv()` further delegates to `cliprdr_process_format_data_response()` which finally delegates to the function assigned as the `ServerFormatDataResponse` handler, which is `guac_rdp_cliprdr_format_data_response()` in our case.
3. If the `ServerFormatDataResponse` handler returns anything but `CHANNEL_RC_OK`, that error code propogates its way back up the chain and results in the thread being killed.

So, we could return an error code here, but it would result in no further CLIPRDR messages being processed. Receiving data in an unexpected format from the RDP server isn't great, and would be a violation of CLIPRDR, but maybe is not so bad that CLIPRDR should be killed for the rest of the session. Returning success here (but logging that things are odd) allows things to continue, assuming copy/paste is otherwise working despite the server misbehaving.

Reply:
Okay, makes sense.  I'm good with it.